docs/topics/db/hooks.txt

```diff
@@ -0,0 +1,99 @@
+======================
+Database execute hooks
+======================
+
+.. versionadded:: 2.0
+		  
+
+In order to help you understand and control the queries issued by your
+code, Django provides a mechanism for installing hooks to be invoked
+around the execution of database queries. Such hooks can be used to
+count queries, measure the time they take, log them, or even prevent
+their execution (e.g. to make sure that no new database queries are
+issued while rendering a template, when you expect it to use only data
+which has been fetched already).
+
+The hooks are defined as context managers -- they are entered before
+the query is executed, and exited after it returns. You can either
+install a contextmanager object, or a callable returning one; if you
+use a callable, it will be called with keyword arguments providing
+information about the query to be executed (currently these are the
+connection, the cursor, the sql text and the parameters).
+
+Installing a hook is itself done in a context manager -- so the hooks
+are temporary and can be made specific to some flow in your code.
+
+Probably the most trivial example of a hook is the execution blocker
+mentioned above.  It could look like this::
+
+    class Blocker:
+        def __enter__(self):
+            raise Exception("No database access allowed here")
+        def __exit__(self, *exc_info):
+            pass
+
+And it would be used in a view to block queries from the template like
+so::
+
+    def my_view(request):
+        context = generate_context(...)
+        template_name = "..."
+        with connection.add_execute_hook(Blocker()):
+            return render(request, template_name, context)
+
+A slightly more complex version would include the connection name in
+the error message::
+
+    @contextmanager
+    def blocker(connection, **kw):
+        raise Exception("Access to databse '{}' blocked here".format(connection.alias))
+        yield
+
+You would use this as ``with connection.add_execute_hook(blocker):``,
+without calling ``blocker`` yourself, so that it is called later with
+the connection object.
+
+A logger of queries could look like this::
+
+    class QueryLogger:
+        def __init__(self):
+            self.queries = []
+        def log(sql, params, *kw):
+            self.queries.append([sql, params])
+            return self
+        def __enter__(self):
+            self.start = time.time()
+        def __exit__(self, *exc_info)
+            duration = time.time() - self.start
+            self.queries[-1].append(duration)
```

Comment:
add spaces between the methods as per normal django style?

Reply:
Yep.